### ISSUE TITLE
Review rng and input requirements in protocols

### ISSUE THREAD

**Issue by amaloz** [boltlabs-inc/tss-ecdsa#290]
Currently, we pass an RNG and `Participant::Input` to all `handle_...` and `gen_...` methods in all `...Participant` protocols. However, in many cases this isn't required. Once we are complete with auditing the protocols, let's revisit this and see if we can simplify the implementations by removing the rng and input requirements where applicable.

Note that while some of this can be done in the audit of the individual protocol, the reason we created a separate issue is because it may affect the way `ProcessOutcome::convert` works, which currently requires both the `rng` and `input` to be provided, so this issue is mainly to flag that as needing a potential redesign based on the requirements of the various protocols.

**Comment by marsella** [comment]
One possible refactor is to make `Input` a field on the `*Participant`, rather than on the `Participant<P>`, so that the `handle...` and `gen...` methods can access directly.

**Comment by marsella** [comment]
[Omar also proposed](https://github.com/boltlabs-inc/tss-ecdsa/pull/302#discussion_r1172049961) making the RNG a field on the `Participant`, which would require more careful guidance for the calling application but would also simplify the handlers.

**Comment by gatoWololo** [comment]
> We have to be a little careful with giving the Participant an RNG to hold because it's easy for the calling application to accidentally clone RNGs and have duplicated state across multiple Participants

I don't fully understand this. What is "duplicate state" and how does it cause issue? Currently, couldn't a user accidentally pass the same rng to multiple `Participant`s as well?

**Comment by marsella** [comment]
> What is "duplicate state" and how does it cause issue? 

If you have an RNG that doesn't re-seed on clone, then you could end up in a situation where multiple threads or participants or whatever object that needed to take ownership of the RNG have synced RNGs -- like if you make N objects at once by cloning the RNG N times, then each object will have the same pseudo-random stream. This breaks the property that we want out of pseudorandomness -- that it's indistinguishable from random. If an attacker learns the state of the RNG of one of those objects, they can use it to learn something about the other objects.

> Currently, couldn't a user accidentally pass the same rng to multiple Participants as well?

Yeah, I didn't think of this. I guess we will have to think about the reseeding situation regardless. 
[Reseeding RNGs exist](https://docs.rs/rand/latest/rand/rngs/adapter/struct.ReseedingRng.html) in Rust, we just have to somehow enforce their use. We might be able to, say, require that the RNG passed to a `Participant` is `CryptoRng + RngCore + ReseedingRng`.

**Comment by marsella** [comment]
Spun off the input requirements into #326 to unblock the issues @Ra1issa mentioned.
